File: app/src/main/java/dating/innovative/gameshowdating/activity/ChatActivity.java
package dating.innovative.gameshowdating.activity;

import android.app.Activity;
import android.content.Intent;
import android.database.sqlite.SQLiteOpenHelper;
import android.os.Bundle;
import android.preference.PreferenceManager;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.View;
import android.widget.Button;
import android.widget.EditText;
import dating.innovative.gameshowdating.R;
import dating.innovative.gameshowdating.model.Message;
import dating.innovative.gameshowdating.util.ChatAdapter;
import dating.innovative.gameshowdating.util.PreferenceManagerClass;

import java.util.ArrayList;
import java.util.Collections;

public class ChatActivity extends Activity {

    RecyclerView messagesView;
    RecyclerView.Adapter messagesAdapter;
    RecyclerView.LayoutManager messagesLayoutManager;
    EditText writeMessageEditText;
    Button sendButton;
    SQLiteHelper dbHelper;

    @Override
    public void onCreate(Bundle savedInstanceState){
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_chat);
        dbHelper = SQLiteHelper.getSqLiteHelperInstance(getApplicationContext());
        final Intent intentFromLast = getIntent();

        final ArrayList<Message> messages = new ArrayList<>();
        /*dbHelper.addMessageToConversationFromSelf(PreferenceManagerClass.getUsername(getApplicationContext()),
                intentFromLast.getStringExtra("username"), "this is a message from self");
        dbHelper.addMessageToConversationFromMatch(PreferenceManagerClass.getUsername(getApplicationContext()),
                intentFromLast.getStringExtra("username"), "this is a message from match");*/

        if(dbHelper.getMessagesForConversation(PreferenceManagerClass.getUsername(getApplicationContext()), intentFromLast.getStringExtra("username")) != null){
            for(int i = 0; i < dbHelper.getMessagesForConversation(PreferenceManagerClass.getUsername(getApplicationContext()),
                    intentFromLast.getStringExtra("username")).size(); i++){
                messages.add(dbHelper.getMessagesForConversation(PreferenceManagerClass.getUsername(getApplicationContext()),
                        intentFromLast.getStringExtra("username")).get(i));
            }
        }
        Collections.reverse(messages);

        messagesView = (RecyclerView) findViewById(R.id.chatRecyclerView);
        messagesLayoutManager = new LinearLayoutManager(this);
        ((LinearLayoutManager) messagesLayoutManager).setReverseLayout(true);
        messagesView.setLayoutManager(messagesLayoutManager);
        messagesAdapter = new ChatAdapter(messages);
        messagesView.setAdapter(messagesAdapter);

        writeMessageEditText = (EditText) findViewById(R.id.chatEditText);
        sendButton = (Button) findViewById(R.id.chatSendMessageButton);
        messagesView.scrollToPosition(0);

        sendButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                if(!writeMessageEditText.getText().toString().isEmpty()){
                    dbHelper.addMessageToConversationFromSelf(PreferenceManagerClass.getUsername(getApplicationContext()),
                            intentFromLast.getStringExtra("username"), writeMessageEditText.getText().toString());
                    messages.add(0,new Message(PreferenceManagerClass.getUsername(getApplicationContext()), intentFromLast.getStringExtra("username"),
                            writeMessageEditText.getText().toString(),null));
                    messagesView.getAdapter().notifyDataSetChanged();
                    messagesView.scrollToPosition(0);
                    writeMessageEditText.setText("");
                }
            }
        });
    }
}

File: app/src/main/java/dating/innovative/gameshowdating/activity/ChatActivity.kt
package dating.innovative.gameshowdating.activity

import android.app.Activity
import android.os.Bundle
import android.support.v7.widget.LinearLayoutManager
import dating.innovative.gameshowdating.R
import dating.innovative.gameshowdating.data.WebSocketHandler
import dating.innovative.gameshowdating.model.Message
import dating.innovative.gameshowdating.util.ChatAdapter
import dating.innovative.gameshowdating.util.PreferenceManagerClass
import dating.innovative.gameshowdating.util.SQLiteHelper
import kotlinx.android.synthetic.main.activity_chat.*

class ChatActivity : Activity() {

    private lateinit var messagesAdapter: ChatAdapter

    public override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_chat)

        val messagesLayoutManager = LinearLayoutManager(this)
        messagesLayoutManager.reverseLayout = true
        chatRecyclerView.layoutManager = messagesLayoutManager
        messagesAdapter = ChatAdapter(listOf<Message>())
        chatRecyclerView.adapter = messagesAdapter
        chatRecyclerView.scrollToPosition(0)

        updateMessages()

        setUpSendMessage()
    }

    private fun setUpSendMessage() {
        val messages = mutableListOf<Message>()
        val dbHelper = SQLiteHelper.getSqLiteHelperInstance(applicationContext)!!
        chatSendMessageButton.setOnClickListener {
            if (!chatEditText.text.toString().isEmpty()) {
                val to = intent.getStringExtra("username")

                dbHelper.addMessageToConversationFromSelf(
                    PreferenceManagerClass.getUsername(applicationContext),
                    intent.getStringExtra("username"), chatEditText.text.toString()
                )

                val message = Message(
                    PreferenceManagerClass.getUsername(applicationContext), intent.getStringExtra("username"),
                    chatEditText.text.toString(), null
                )
                WebSocketHandler.instance.sendChatMessage(to, message.messageFromSelf, System.currentTimeMillis()) {
                    updateMessages()
                }
                messages.add(0, message)
                messagesAdapter.notifyDataSetChanged()
                chatRecyclerView.scrollToPosition(0)
                chatEditText.setText("")
            }
        }
    }

    private fun updateMessages() {
        val toUser = intent.getStringExtra("username")

        WebSocketHandler.instance.getMessages(PreferenceManagerClass.getUsername(this)) { messageMap ->
            val messageList = messageMap[toUser]?: listOf()
            messagesAdapter.messages = messageList
            runOnUiThread {
                messagesAdapter.notifyDataSetChanged()
            }
        }
    }
}